### ISSUE TITLE
YOLOv5 incapable of detecting very underrepresented class

### ISSUE THREAD

**Issue by axelpuyo** [ultralytics/yolov5#11965]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello,

I have a dataset of images where I'm trying to detect 2 types of objects. One of them is extremely overrepresented, and the other one underrepresented. For reference, in an image, I have about 350 objects of class 1, and at best 10 of class 2 (often 1 or none).
In the entire dataset, I have 51k counts of class 1, and 1k of class 2. At training time, YOLOv5 gives up on detecting class 2 and overfits to class 1.

Unfortunately, due to reasons pertaining to our domain of application, we are forced to detect both simultaneously (not possible to use YOLO as an anomaly detector for class 2).

I have scoured issues and discussions to no avail. Thus my question is the following: is there a way to force YOLO to fit to class 2?
I have attempted passing the argument `--image-weights`, tinkering `fl_gamma` of the focal loss, and passing class weights to `pos_weight` in `BCEWithLogitsLoss` to no avail.

Thank you for time.

**Comment by glenn-jocher** [comment]
@axelpuyo hi,

I understand your concern about YOLOv5 not detecting the underrepresented class effectively. YOLOv5 relies on a balanced dataset for optimal performance, and training on imbalanced data can be challenging. 

However, there are a few things you can try to improve the detection of the underrepresented class:

1. **Data Augmentation**: Apply augmentation techniques like flipping, scaling, rotation, or adding noise to the images of the underrepresented class. This can help increase the variability of the data and provide additional examples for training.

2. **Data Sampling**: Consider oversampling the underrepresented class or using techniques like Synthetic Minority Over-sampling Technique (SMOTE) to generate synthetic examples. This can help balance the representation of both classes during training.

3. **Class Weights**: Assign higher weights to the samples of the underrepresented class during training. This can help in preventing the model from favoring the dominant class during optimization. You can adjust the `pos_weight` parameter in the `torch.nn.BCEWithLogitsLoss` function to accomplish this.

4. **Model Architecture**: Experiment with different model architectures or configurations that are known to perform well on imbalanced datasets. This may involve adjusting the network structure or introducing additional components like attention mechanisms.

Keep in mind that achieving strong detection performance on highly imbalanced datasets can be challenging. It's important to carefully evaluate the results and consider the trade-offs between false positives and false negatives.

I hope these suggestions help. Feel free to reach out if you have any further questions or need additional assistance.

Best regards.

**Comment by axelpuyo** [comment]
> @axelpuyo hi,
> 
> I understand your concern about YOLOv5 not detecting the underrepresented class effectively. YOLOv5 relies on a balanced dataset for optimal performance, and training on imbalanced data can be challenging.
> 
> However, there are a few things you can try to improve the detection of the underrepresented class:
> 
> 1. **Data Augmentation**: Apply augmentation techniques like flipping, scaling, rotation, or adding noise to the images of the underrepresented class. This can help increase the variability of the data and provide additional examples for training.
> 2. **Data Sampling**: Consider oversampling the underrepresented class or using techniques like Synthetic Minority Over-sampling Technique (SMOTE) to generate synthetic examples. This can help balance the representation of both classes during training.
> 3. **Class Weights**: Assign higher weights to the samples of the underrepresented class during training. This can help in preventing the model from favoring the dominant class during optimization. You can adjust the `pos_weight` parameter in the `torch.nn.BCEWithLogitsLoss` function to accomplish this.
> 4. **Model Architecture**: Experiment with different model architectures or configurations that are known to perform well on imbalanced datasets. This may involve adjusting the network structure or introducing additional components like attention mechanisms.
> 
> Keep in mind that achieving strong detection performance on highly imbalanced datasets can be challenging. It's important to carefully evaluate the results and consider the trade-offs between false positives and false negatives.
> 
> I hope these suggestions help. Feel free to reach out if you have any further questions or need additional assistance.
> 
> Best regards.

Hello @glenn-jocher, thank you for your quick response.

Unfortunately as I mentioned, the best images I have (wrt to the underrepresented class) have at best a ratio of 1:50 which is still very bad. So even with data augmentation, I do not think a healthy ratio is achievable. As for SMOTE, I thought YOLO already did this when you pass `--image-weights` argument in training through the use of `WeightedRandomSampler`. Additionally, I do pass the pre-computed class_weights to `BCEWithLogitsLoss` through the use of `cls_pw` in my `hyps.yaml` file. Even with the combination of all these approaches, the best performance I achieve is 90% precision on the overrepresented class, and 30% on the underrepresented class.

I will look into adding attention mechanisms to the architecture of the model.
If this doesn't work, is there another version of YOLO or any other Ultralytics Object Detection model which is better suited to deal with severely unbalanced datasets?

Thank you for time again.

**Comment by glenn-jocher** [comment]
Hello @axelpuyo,

I understand your concern about the severely unbalanced dataset and the challenges it presents for YOLOv5. It seems you have already tried various approaches like data augmentation, SMOTE, and adjusting class weights.

Regarding your question about other versions of YOLO or Ultralytics object detection models that may be better suited for severely unbalanced datasets, I recommend looking into RetinaNet, which is another popular object detection architecture. RetinaNet is known to handle imbalanced datasets more effectively due to its focal loss, which assigns higher weights to hard examples (which often correspond to the underrepresented class).

You can find an implementation of RetinaNet in the [Ultralytics det](https://github.com/ultralytics/yolov5/tree/master/det) repository. This repository contains the code for training and evaluating different object detection models, including RetinaNet.

I hope this information helps. Let me know if you have any further questions or need additional assistance.

**Comment by tejasri19** [comment]
> Hello @axelpuyo,
> 
> I understand your concern about the severely unbalanced dataset and the challenges it presents for YOLOv5. It seems you have already tried various approaches like data augmentation, SMOTE, and adjusting class weights.
> 
> Regarding your question about other versions of YOLO or Ultralytics object detection models that may be better suited for severely unbalanced datasets, I recommend looking into RetinaNet, which is another popular object detection architecture. RetinaNet is known to handle imbalanced datasets more effectively due to its focal loss, which assigns higher weights to hard examples (which often correspond to the underrepresented class).
> 
> You can find an implementation of RetinaNet in the [Ultralytics det](https://github.com/ultralytics/yolov5/tree/master/det) repository. This repository contains the code for training and evaluating different object detection models, including RetinaNet.
> 
> I hope this information helps. Let me know if you have any further questions or need additional assistance.

The link provided for retinanet models is broken. Please share the updated one. @glenn-jocher

**Comment by glenn-jocher** [comment]
Hello @tejasri19,

Thank you for pointing out the broken link. I apologize for the inconvenience.

For handling severely unbalanced datasets, you might want to explore models that incorporate focal loss, such as RetinaNet. Unfortunately, Ultralytics' YOLOv5 repository does not directly include RetinaNet implementations. However, you can find RetinaNet implementations in other repositories, such as the [torchvision](https://github.com/pytorch/vision) library, which provides a robust implementation of RetinaNet.

Here's a quick guide on how to get started with RetinaNet using torchvision:

1. **Install torchvision**:
   ```bash
   pip install torchvision
   ```

2. **Load a pre-trained RetinaNet model**:
   ```python
   import torchvision
   model = torchvision.models.detection.retinanet_resnet50_fpn(pretrained=True)
   ```

3. **Fine-tune the model on your dataset**:
   You can follow the [torchvision object detection tutorial](https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html) to fine-tune the model on your custom dataset.

Additionally, if you are looking for more advanced techniques to handle class imbalance within YOLOv5, you might consider:

- **Custom Loss Functions**: Implementing a custom loss function that incorporates focal loss directly into YOLOv5.
- **Transfer Learning**: Using transfer learning to leverage pre-trained weights and fine-tuning the model on your dataset with a focus on the underrepresented class.

For more detailed guidance on these techniques, you can refer to the [Ultralytics YOLOv5 documentation](https://docs.ultralytics.com/yolov5/).

I hope this helps! If you have any further questions or need additional assistance, feel free to ask. 😊
